File: packages/ui/Sources/UI/Toast/ToastController.swift
import Combine

/// App-wide transient-message state. Callers invoke `show`; `ToastHost` renders it via AlertToast.
@MainActor
public final class ToastController: ObservableObject {
  @Published public var isPresenting = false

  public private(set) var message = ""
  public private(set) var tone: ToastTone = .danger

  public init() {}

  public func show(_ message: String, tone: ToastTone = .danger) {
    self.message = message
    self.tone = tone
    self.isPresenting = true
  }
}

File: packages/ui/Java/com/variance/nearby/ui/ToastController.kt
package com.variance.nearby.ui

import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow

data class ToastEvent(
    val message: String,
    val tone: ToastTone,
)

/**
 * App-wide transient-message bus. Callers invoke [show]; `ToastHost` collects [events] and renders
 * them through compose-sonner. Decouples message producers from the Compose toaster state.
 */
class ToastController {
    private val mutableEvents = MutableSharedFlow<ToastEvent>(extraBufferCapacity = 8)
    val events: SharedFlow<ToastEvent> = mutableEvents.asSharedFlow()

    fun show(
        message: String,
        tone: ToastTone = ToastTone.NEUTRAL,
    ) {
        mutableEvents.tryEmit(ToastEvent(message, tone))
    }
}
